src/vt_flash/newton.jl

```diff
@@ -148,11 +170,56 @@ function newton(
         @debug "newton" i repr(δx) norm_step=norm(δx, 2) α αmax repr(x) fval fcalls norm_gprev=norm(∇f!(similar(x), x - α*δx)) norm_gnew=norm(∇) prod(diag(hess.U))
 
         # check for convergence
-        norm(∇, 2) ≤ gtol && return NewtonResult(true, x, i, totfcalls)
+        convcond(x, ∇) && return NewtonResult(true, x, i, totfcalls)
+        !isnan(gtol) && norm(∇, 2) ≤ gtol && return NewtonResult(true, x, i, totfcalls)
     end
     return NewtonResult(false, x, maxiter, totfcalls)
 end
 
+function __convergence_closure(
+    state1::AbstractVTFlashState,
+    mix::BrusilovskyEoSMixture,
+    nmolb::AbstractVector,
+    volumeb::Real,
+    RT::Real;
+    chemrtol::Real,
+    pressrtol::Real,
+    buf::BrusilovskyThermoBuffer=thermo_buffer(mix),
+)
+    state1x = value(state1)
+
+    g1 = Vector{Float64}(undef, ncomponents(mix) + 1)
+    g2 = similar(g1)
+    diff = similar(g1)
+    nmol2 = similar(g1, Float64, ncomponents(mix))
+
+    function convcond(x::AbstractVector, g::AbstractVector)
+        state1x .= x
+        nmol1, vol1 = nmolvol(state1, nmolb, volumeb)
+        g1 = CubicEoS.nvtgradient!(g1, mix, nmol1, vol1, RT; buf=buf)
+
+        @. nmol2 = nmolb - nmol1
+        vol2 = volumeb - vol1
+        g2 = CubicEoS.nvtgradient!(g2, mix, nmol2, vol2, RT; buf=buf)
+
+        #=
+                   [μ₁' - μ₁'', ..., μₙ' - μₙ''; -P' + P'']ᵀ
+            diff = ----------------------------------------
+                                    RT
+        =#
+        @. diff = g1 - g2
+
+        nrmchem = let chem1 = (@view g1[1:end-1]), diffchem = (@view diff[1:end-1])
+            abs.(diffchem) ./ abs.(chem1) |> maximum
+        end
+        nrmpress = abs(diff[end] / g1[end])
+
+        return nrmchem ≤ chemrtol && nrmpress ≤ pressrtol
```

Comment:
Draft of convergence criterion. The criterion is

```
1) max_i | chempot'_i - chempot''_i | / | chempot'_i | <= chemrtol
2)       | pressure' - pressure'' | / | pressure' | <= pressrtol
```

If pressure of (first) phase `pressure'` is small, we may consider a mixed approach similar to [FNC](http://tobydriscoll.net/fnc-julia/localapprox/adaptive.html#divide-and-conquer) (see 5.7.4 and line 24 of `intadapt`).

So

```
abs(pdiff) <= presstol * (1 + pressure')
```

Reply:
@vvpisarev

Reply:
The mentioned approach for the tolerances seems as a special case from http://dx.doi.org/10.1016/j.fluid.2015.07.035 (Equation 60)

<img width="614" alt="image" src="https://user-images.githubusercontent.com/11893064/153884655-6583b39c-fd23-4e61-83be-43c77c7eba15.png">

Reply:
After discussion, the convergence criterion for chemical potentials is | mu'_i - mu''_i | / RT for all i.